### PR TITLE
TUNIP

### PR DESCRIPTION
**Pull request by junrushao1994**
Infra challenges:
- [x] Concise scoping
- [x] Special constant: T/R
- [x] IRScope-based dispatch
- [x] Undefined variables => retrieve a frame which is either TIRFrame, or VarDefFrame
- [x] Sugaring operator calls => op attributes
- [x] Variables without name, e.g. buffer.data

### CODE REVIEW

File: src/script/printer/frame.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#include "./symbol_table.h"
+
+namespace tvm {
+namespace script {
+namespace printer {
+
+class FrameNode : public Object {
+ public:
+  Array<ObjectRef> objs;
+  SymbolTable sym;
```
Comment (junrushao1994) [inline src/script/printer/frame.h:33]:
```suggestion
  SymbolTableNode *sym;
```

File: src/script/printer/doc.h [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+class ForDoc : public StmtDoc {
+ public:
+  explicit ForDoc(ExprDoc lhs, ExprDoc rhs, Array<StmtDoc> body);
+  TVM_DEFINE_NOTNULLABLE_OBJECT_REF_METHODS(ForDoc, StmtDoc, ForDocNode);
+};
+
+class ScopeDocNode : public StmtDocNode {
+ public:
+  ExprDoc lhs{nullptr};
+  ExprDoc rhs{nullptr};
```
Comment (yelite) [inline src/script/printer/doc.h:413]: Should the rhs be Optional<ExprDoc>?

File: src/script/printer/doc.h [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+
+class ScopeDoc : public StmtDoc {
+ public:
+  explicit ScopeDoc(ExprDoc lhs, ExprDoc rhs, Array<StmtDoc> body);
+  TVM_DEFINE_NOTNULLABLE_OBJECT_REF_METHODS(ScopeDoc, StmtDoc, ScopeDocNode);
+};
+
+class ExprStmtDocNode : public StmtDocNode {
+ public:
+  Optional<ExprDoc> expr{NullOpt};
```
[the remark this answers is not in the document]
Comment (yelite) [reply src/script/printer/doc.h:435]: Create a new node to represent block comment, and make this required

File: src/script/printer/doc.h (hunk 2 above)
Comment (junrushao1994) [reply src/script/printer/doc.h:413]: Yes for LHS

**Pull request closed without merging**